File: src/main/java/database/finance/ExpensesTable.java
package database.finance;

import com.koenig.commonModel.finance.BookkeepingEntry;
import com.koenig.commonModel.finance.Expenses;
import database.NamedParameterStatement;
import org.joda.time.DateTime;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Arrays;
import java.util.Collection;

public class ExpensesTable extends BookkeepingTable<Expenses> {

    public static final String NAME = "expenses_table";
    private static final String DATE = "date";
    private static final String STANDING_ORDER = "standing_order";

    public ExpensesTable(Connection connection) {
        super(connection);
    }

    @Override
    protected Expenses getBookkeepingItem(BookkeepingEntry entry, ResultSet rs) throws SQLException {
        DateTime date = getDateTime(rs, DATE);
        String standingOrder = rs.getString(STANDING_ORDER);
        return new Expenses(entry, date, standingOrder);
    }

    @Override
    protected String getBookkeepingTableSpecificCreateStatement() {
        return "," + DATE + " TEXT, " + STANDING_ORDER + " LONG";
    }

    @Override
    protected void setBookkeepingItem(NamedParameterStatement ps, Expenses item) throws SQLException {
        setDateTime(ps, DATE, item.getDate());
        ps.setString(STANDING_ORDER, item.getStandingOrder());
    }

    @Override
    protected Collection<? extends String> getBookkeepingColumnNames() {
        return Arrays.asList(DATE, STANDING_ORDER);
    }

    @Override
    public String getTableName() {
        return NAME;
    }
}

File: src/main/java/database/finance/ExpensesTable.kt
package database.finance

import com.koenig.commonModel.finance.BookkeepingEntry
import com.koenig.commonModel.finance.Expenses
import database.NamedParameterStatement
import org.joda.time.LocalDate
import java.sql.Connection
import java.sql.ResultSet
import java.sql.SQLException
import java.util.*

class ExpensesTable(connection: Connection) : BookkeepingTable<Expenses>(connection) {

    override val bookkeepingTableSpecificCreateStatement: String
        get() = ",$DAY LONG, $STANDING_ORDER TEXT, $COMPENSATION INT"

    override val bookkeepingColumnNames: Collection<String>
        get() = Arrays.asList(DAY, STANDING_ORDER, COMPENSATION)

    override val tableName: String
        get() = NAME

    @Throws(SQLException::class)
    override fun getBookkeepingItem(entry: BookkeepingEntry, rs: ResultSet): Expenses {
        val date = getLocalDate(rs, DAY)
        val standingOrder = rs.getString(STANDING_ORDER)
        val isCompensation = rs.getBoolean(COMPENSATION)
        return Expenses(entry, date, standingOrder, isCompensation)
    }


    @Throws(SQLException::class)
    override fun setBookkeepingItem(ps: NamedParameterStatement, item: Expenses) {
        ps.setLocalDate(DAY, item.day)
        ps.setString(STANDING_ORDER, item.standingOrder)
        setBool(ps, COMPENSATION, item.isCompensation)
    }

    companion object {
        val NAME = "expenses_table"
        private const val DAY = "date"
        private const val STANDING_ORDER = "standing_order"
        private const val COMPENSATION = "compensation"
    }

    val compensations: Map<LocalDate, Expenses>
        get() {
            val list = getWith(getNamedParameter(COMPENSATION), { it.setBoolean(COMPENSATION, true) })
            val map = mutableMapOf<LocalDate, Expenses>()
            list.forEach { map[it.item.day] = it.item }
            return map
        }


}
